vllm/v1/sample/logits_processor/__init__.py

```diff
@@ -177,9 +181,125 @@ def build_logitsprocs(
 
 
+V0WrapperState = tuple[LogitsProcessorV0, list[int], list[int]]
+
+
+class LogitsProcessorV0Wrapper(LogitsProcessor):
+    """Wrapper for vLLM V0-style `Callable` logits processors
+    
+    To wrap a specific `Callable` logits processor, 
+    * Implement `self.is_argmax_invariant()` base-class method
+    * Implement `self._new_logits_processor()`
+    * Implement `self._nargs` property
+    
+    """
+
+    def __init__(self, vllm_config: VllmConfig, device: torch.device,
+                 is_pin_memory: bool):
+        # For each request that is assigned a key in self.req_info, store an
+        # optional V0-style logits processor, prompt token ids, and previously
+        # decoded token ids
+        self.req_info: dict[int, V0WrapperState] = {}
+
+        # Preserve config info for use when constructing V0-style logits
+        # processors
+        self.vllm_config = vllm_config
+        self.device = device
+        self.is_pin_memory = is_pin_memory
+
+        # V0 logits processors can have 2 or 3 arguments
+        assert (nargs := self._nargs) in {2, 3}
+        self.nargs = nargs
+
+    @property
+    @abstractmethod
+    def _nargs(self) -> int:
+        """Returns the number of arguments supported by this V0 logits
+        processor. Must be 2 or 3"""
+        raise NotImplementedError()
+
+    @abstractmethod
+    def _new_logits_processor(
+        self,
+        params: SamplingParams,
+    ) -> Optional[LogitsProcessorV0]:
+        """Consume request info; return a V0-style logits processor.
+
+        Implementer can assume that the vLLM engine config, the hardware
+        accelerator device, and an indication of whether pin memory is
+        available can respectively be found in self.vllm_config, self.device,
+        and self.is_pin_memory
+
+        Return None if logits processor does not need to be applied to request
+
+        Args:
+          params: request sampling params
+
+        Returns:
+          None if logits processor should not be applied to request; otherwise
+          returns a `LogitsProcessorV0` instance
+        
+        """
+        raise NotImplementedError()
+
+    def _new_state(
+        self,
+        params: SamplingParams,
+        prompt_tok_ids: list[int],
+        out_tok_ids: list[int],
+    ) -> Optional[V0WrapperState]:
+        """Return state representation for new request
+
+        Returns None if logits processor is not applicable to request
+
+        Args:
+          params: request sampling params
+          prompt_tok_ids: request prompt token ids
+          out_tok_ids: decoded tokens so far for this request
+
+        Returns:
+          (new logits processor, prompt token ids,
+           previous decoded tokens) or None
+        
+        """
+        if (v0_lp := self._new_logits_processor(params)):
+            return (
+                v0_lp,
+                prompt_tok_ids,
+                out_tok_ids,
+            )
+        return None
```

Comment:
Can we use the `partial` approach I suggested here: https://github.com/njhill/vllm/commit/2b205b31fe02c0a4a2b190e66042398cc3e67111#diff-6278b323b85096bd74b396634d6f6edc78a369b4ce50d52dc8f3203ab0d3ada8R45-R52

Then the `apply` method is much simpler. Also by introspecting the returned LP we don't need them to explicitly provide the number of args.

And the `_new_state` method then also disappears